pkg/ipam-node/handlers/allocate.go

```diff
@@ -0,0 +1,159 @@
+		return nil, err
+	}
+	resp := &nodev1.AllocateResponse{}
+	for _, r := range result {
+		resp.Allocations = append(resp.Allocations, &nodev1.AllocationInfo{
+			Pool:    r.Pool,
+			Ip:      r.Address.String(),
+			Gateway: r.Gateway.String(),
+		})
+	}
+	return resp, nil
+}
+
+// PoolAlloc container which store Pool name and allocation
+type PoolAlloc struct {
+	Pool string
+	*current.IPConfig
+}
+
+func (h *Handlers) allocate(reqLog logr.Logger,
+	store storePkg.Store, params *nodev1.IPAMParameters) ([]PoolAlloc, error) {
+	var err error
+	result := make([]PoolAlloc, 0, len(params.Pools))
+	for _, pool := range params.Pools {
+		var alloc PoolAlloc
+		alloc, err = h.allocateInPool(pool, reqLog, store, params)
+		if err != nil {
+			break
+		}
+		result = append(result, alloc)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (h *Handlers) allocateInPool(pool string, reqLog logr.Logger,
+	store storePkg.Store, params *nodev1.IPAMParameters) (PoolAlloc, error) {
+	poolLog := reqLog.WithValues("pool", pool)
+
+	poolCfg := h.poolManager.GetPoolByName(pool)
+	if poolCfg == nil {
+		return PoolAlloc{}, status.Errorf(codes.NotFound, "configuration for pool %s not found", pool)
+	}
+	rangeStart := net.ParseIP(poolCfg.StartIP)
+	if rangeStart == nil {
+		return PoolAlloc{}, poolCfgError(poolLog, pool, "invalid rangeStart")
+	}
+	rangeEnd := net.ParseIP(poolCfg.EndIP)
+	if rangeEnd == nil {
+		return PoolAlloc{}, poolCfgError(poolLog, pool, "invalid rangeEnd")
+	}
+	_, subnet, err := net.ParseCIDR(poolCfg.Subnet)
+	if err != nil || subnet == nil || subnet.IP == nil || subnet.Mask == nil {
+		return PoolAlloc{}, poolCfgError(poolLog, pool, "invalid subnet")
+	}
+	gateway := net.ParseIP(poolCfg.Gateway)
+	if gateway == nil {
+		return PoolAlloc{}, poolCfgError(poolLog, pool, "invalid gateway")
+	}
+	rangeSet := &allocator.RangeSet{allocator.Range{
+		RangeStart: rangeStart,
+		RangeEnd:   rangeEnd,
+		Subnet:     cniTypes.IPNet(*subnet),
+		Gateway:    gateway,
+	}}
+	if err := rangeSet.Canonicalize(); err != nil {
+		return PoolAlloc{}, poolCfgError(poolLog, pool,
+			fmt.Sprintf("invalid range config: %s", err.Error()))
+	}
+	alloc := h.getAllocFunc(rangeSet, pool, store)
+	allocMeta := types.ReservationMetadata{
+		CreateTime:         time.Now().Format(time.RFC3339Nano),
+		PoolConfigSnapshot: poolCfg.String(),
+	}
+	if params.Metadata != nil {
+		allocMeta.PodUUID = params.Metadata.K8SPodUid
+		allocMeta.PodName = params.Metadata.K8SPodName
+		allocMeta.PodNamespace = params.Metadata.K8SPodNamespace
+		allocMeta.DeviceID = params.Metadata.DeviceId
+	}
+	result, err := alloc.Allocate(params.CniContainerid, params.CniIfname, allocMeta)
+	if err != nil {
+		poolLog.Error(err, "failed to allocate IP address")
+		if errors.Is(err, storePkg.ErrReservationAlreadyExist) {
+			return PoolAlloc{}, status.Errorf(codes.AlreadyExists, "IP address in pool %s is already allocated", pool)
```

Comment:
i dont think it means that IP is already allocated but rather that this container + interface already has an IP allocated from the pool.

Reply:
I changed the error message